core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -737,18 +737,48 @@ public long getFileId(AlluxioURI path) throws AccessControlException, Unavailabl
 
   private long getFileIdInternal(AlluxioURI path, boolean checkPermission)
       throws AccessControlException, UnavailableException {
-    try (RpcContext rpcContext = createRpcContext();
-         LockedInodePath inodePath = mInodeTree.lockInodePath(path, LockPattern.READ)) {
-      if (checkPermission) {
-        mPermissionChecker.checkPermission(Mode.Bits.READ, inodePath);
+    try (RpcContext rpcContext = createRpcContext()) {
+      /*
+      In order to prevent locking twice on RPCs where metadata does _not_ need to be loaded, we use
+      a two-phase scheme as an optimization to prevent the extra lock. loadMetadataIfNotExists
+      requires a lock on the tree to determine if the path should be loaded before executing. To
+      prevent the extra lock, we execute the RPC as normal and use a conditional check in the
+      main body of the function to determine whether control flow should be shifted out of the
+      RPC logic and back to the loadMetadataIfNotExists function.
+
+      If loadMetadataIfNotExists runs, then the next pass into the main logic body should
+      continue as normal. This may present a slight decrease in performance for newly-loaded
+      metadata, but it is better than affecting the most common case where metadata is not being
+      loaded.
+       */
+      LoadMetadataContext lmCtx = LoadMetadataContext.mergeFrom(
+          LoadMetadataPOptions.newBuilder().setCreateAncestors(true));
+      boolean run = true;
+      boolean loadMetadata = false;
+      while (run) {
```

Comment:
Is there any chance this could run for more than 2 iterations?

Reply:
Do you see a case where it could? 

I stepped through the logic myself to iron out as many errors as possible when writing it and when doing the tests.

For each of the implementations there are two flags 
- `run`
- `loadMetadata`


In the main body of the loop `run` is always immediately set `false` after entering the loop and is only set again within a conditional that contains `if (!loadMetadata)`

Within the conditional of `if (!loadMetadata)` we set `run` to `true`, but also set the `loadMetadata` flag to true, which means that we shouldn't enter the same conditional the 2nd time around to be able to set `run` to true again - preventing more than 2 iterations.

Reply:
If you have a better idea of how you would implement this logic I'm all ears. I'm sure there are ways it can be improved. This was the simplest thing I came up with.

Reply:
Yeah, after looking at it, it should be fine. The important part is that we have to make sure run is set to false in the beginning, and setting `run=true` can only happen when `loadMetadata == false`.